src/Voron/Data/CompactTrees/CompactTree.cs

```diff
@@ -34,59 +34,518 @@ private struct IteratorCursorState
+                _internalKeyCache2 = null;
             }
 
-            public override string ToString()
+            return current;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ReleaseKey(EncodedKey key)
+        {
+            if (_internalKeyCache1 != null && _internalKeyCache2 != null)
+            {
+                key.Dispose();
+                return;
+            }
+
+            if (_internalKeyCache2 != null)
+            {
+                _internalKeyCache1 = key;
+            }
+            else
+            {
+                _internalKeyCache2 = key;
+            }
+        }
+
+        public readonly struct EncodedKeyScope : IDisposable
+        {
+            public readonly EncodedKey Key;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public EncodedKeyScope(CompactTree tree)
+            {
+                Key = tree.AcquireKey();
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void Dispose()
+            {
+                Key.Owner.ReleaseKey(Key);
+            }
+        }
+
+        public class EncodedKey : IDisposable
+        {
+            public readonly CompactTree Owner;
+
+            private ByteString _storage;
+            private ByteStringContext<ByteStringMemoryCache>.InternalScope _storageScope;
+            
+            // The storage data will be used in an arena fashion. If there is no enough, we just create a bigger one and
+            // copy the content back. 
+            private byte* _currentPtr;
+            private byte* _currentEndPtr;
+
+            // The decoded key pointer points toward the actual decoded key (if available). If not we will take the current
+            // dictionary and just decode it into the storage. 
+            private int _decodedKeyIdx;
+
+            // The mapping dictionary works like glorified cache but it is really more like a hash-table.
+            // The base idea is to access directly the encoded key representation already stored in this
+            // key auxiliary storage in order to reuse and avoid redundant work. 
+            // In true DOS fashion, 64 values ought to be enough for anybody.
+            private struct KeyMapping
+            {
+                // If the dictionary is 0 then the dictionary is not present in this key.
+                public long Dictionary;
+
+                // If we have the encoded version of the key available already, the KeyIndex will be >0 and the Dictionary Valid. 
+                public int KeyIndex;
+            }
+
+            private int _currentKeyIdx;
+
+            private const int MappingTableSize = 64;
+            private const int MappingTableMask = MappingTableSize - 1;
+
+            private readonly long[] _keyMappingDictionary = new long[MappingTableSize];
+            private readonly int[] _keyMappingIndex = new int[MappingTableSize];
```

Comment:
Those are used only _during_ the searching process, right?
Can we make the thread static, then? Would save some allocations.

Reply:
Yes. Haven't look yet at what allocation avoidance technique would be proper to use here.